### DISCUSSION TITLE
run_sync usage with async_scoped_session

### DISCUSSION THREAD

**Discussion by pankajkumar229** [sqlalchemy/sqlalchemy#9166] [Usage Questions] [answered]
In my program I have been using async_scoped_session all over and not using the engine directly.

Many of the ORM backrefs do not seem to be loaded. run_sync works but the code goes through directly with the engine:

This code seems to work:

```
+    async def load_courses_to_user(self, user):
+        def load_assignments(session) :
+            for user_assignment_association in user.assignments:
+               user_assignment_association.assignment.course.instructor;
+
+        async with self.engine.begin() as conn:
+                await conn.run_sync(load_assignments)
```
Is this safe to use with async_scoped_session?

**Comment by CaselIT** [comment, marked as answer]
Hi,

Thar should work but you are probably better off using the rum symc of the session.

It seems that run_sync is missing from the async_scoped_sessiom. Is should not be hard to add. For now you caouls use `your_async_scoped_session().run_sync`

**Reply by pankajkumar229** [reply]
Hi, I changed to 

```
+    async def load_courses_to_user(self, user):
+
+         def load_assignments(session) :
+             for user_assignment_association in user.assignments:
+                 user_assignment_association.assignment.course.instructor;
+             for course_user_association in user.courses:
+                 course_user_association.course.instructor;
+         self.db().run_sync(load_assignments)
```

self.db is an instance of async_scoped_session. but load_assignments does not seem to get called at all.

**Reply by CaselIT** [reply]
you need to await the run_sync.

Try doing: `await self.db().run_sync(load_assignments)`

**Reply by pankajkumar229** [reply]
Thanks!! I ended up creating a helper function that I call before accessing anything that is not loaded:

```
+    async def load_attributes_to_object(self, main_obj, attrs):
+         next_obj = main_obj
+         def load_attributes_sync(session) :
+             nonlocal next_obj, attrs
+             for x in attrs:
+                 next_obj = getattr(next_obj, x)
+
+         await self.db().run_sync(load_attributes_sync)
+
```

Could something like this not be called directly by SQLAlchemy perhaps only if the user sets a configuration?

**Reply by CaselIT** [reply]
At the moment sqlalchemy supports settings eager loaders defined globally in the relationship or on a per query basis using options.

If you already know what relationships will be required at query time is more efficient to use eager loading. More details at https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html
